### PR TITLE
feat: incremental route radixtree update

### PR DESCRIPTION
**Pull request by ranxuxin001**
refer to #9259

This feature also depends on the extended source codes in the lua-resty-radixtree library. https://github.com/api7/lua-resty-radixtree/pull/128 

pressure test:
environment: centos 7   8core  32G mem

wrk -c1000 -d60s -t8 --latency --timeout=15s  http://www.aaabbb1122.com:9080/123

patched version:
image

image

elapsed time to change a route:
image

original version:
image

image

elapsed time to change a route:
image

### CODE REVIEW

**Comment by lingliy** [comment]
I also really need this incremental update feature, there are two puzzles
1. According to the routing semantics of apisix, when route or service changes, the route is rebuilt, but service changes are not observed. How is the incremental update implemented
2.radixtree_host_uri.lua, radixtree_uri_with_parameter.lua, these two related incremental updates do not seem to be implemented
Finally, thank you very much for the implementation of feature

**Comment by ranxuxin001** [comment]
> I also really need this incremental update feature, there are two puzzles
> 
> 1. According to the routing semantics of apisix, when route or service changes, the route is rebuilt, but service changes are not observed. How is the incremental update implemented
>    2.radixtree_host_uri.lua, radixtree_uri_with_parameter.lua, these two related incremental updates do not seem to be implemented
>    Finally, thank you very much for the implementation of feature

Thank you to send request. I wrote this patch based on apisix 2.9. The radixtree_uri.lua runs as default. So I mainly wrote codes for this strategy. Later I will wrote codes based on radixtree_host_uri.lua. 

I deleted codes rebuilding whote tree when version changed because the modifying routes codes are moved in the filter() called by sync_data().

**Comment by ranxuxin001** [comment]
> > I also really need this incremental update feature, there are two puzzles
> > 
> > 1. According to the routing semantics of apisix, when route or service changes, the route is rebuilt, but service changes are not observed. How is the incremental update implemented
> >    2.radixtree_host_uri.lua, radixtree_uri_with_parameter.lua, these two related incremental updates do not seem to be implemented
> >    Finally, thank you very much for the implementation of feature
> 
> Thank you to send request. I wrote this patch based on apisix 2.9. The radixtree_uri.lua runs as default. So I mainly wrote codes for this strategy. Later I will wrote codes based on radixtree_host_uri.lua as the default strategy of apisix 3.3.
> 
> I deleted codes rebuilding whote tree when version changed because the modifying routes codes are moved in the filter() called by sync_data().

**Comment by lingliy** [comment]
> I also really need this incremental update feature, there are two puzzles
> 
> 1. According to the routing semantics of apisix, when route or service changes, the route is rebuilt, but service changes are not observed. How is the incremental update implemented
>    2.radixtree_host_uri.lua, radixtree_uri_with_parameter.lua, these two related incremental updates do not seem to be implemented
>    Finally, thank you very much for the implementation of feature

I see. I read the semantics of version 2.9, and routes will not be rebuilt after service changes. Later versions will appear

File: apisix/router.lua [hunk 1]
```diff
@@ -22,34 +22,147 @@ local plugin_checker = require("apisix.plugin").plugin_checker
...
+local json         = require("apisix.core.json")
 
 local _M = {version = 0.3}
 
+local function short_key(self, str)
+    return sub_str(str, #self.key + 2)
+end
 
-local function filter(route)
+local function filter(route, pre_route_obj, size)
```
Comment (kingluo) [inline apisix/router.lua:35]: The function prototype is too obscure, `pre_route_obj` may contain different types. It's better to split the radix tree reconstruction into another function, e.g. `filter_all_routes`.

File: apisix/router.lua [hunk 2]
```diff
@@ -70,6 +183,7 @@ local function attach_http_router_common_methods(http_router)
     end
 end
 
+local uri_routes = {}
```
Comment (kingluo) [inline apisix/router.lua:186]: dead code?

File: apisix/router.lua [hunk 3]
```diff
@@ -84,6 +198,7 @@ function _M.http_init_worker()
     local router_http = require("apisix.http.router." .. router_http_name)
     attach_http_router_common_methods(router_http)
     router_http.init_worker(filter)
+
```
Comment (kingluo) [inline apisix/router.lua:201]: extra blank line?

File: apisix/router.lua [hunk 4]
```diff
@@ -22,34 +22,147 @@ local plugin_checker = require("apisix.plugin").plugin_checker
...
 
+local function short_key(self, str)
+    return sub_str(str, #self.key + 2)
+end
 
-local function filter(route)
+local function filter(route, pre_route_obj, size)
     route.orig_modifiedIndex = route.modifiedIndex
-    route.update_count = 0
+    route.update_count = 0
```
Comment (kingluo) [inline apisix/router.lua:37]: extra space.

File: apisix/http/router/radixtree_uri.lua [hunk 5]
```diff
@@ -17,33 +17,15 @@
...
 
 
 local _M = {version = 0.2}
 
 
-    local uri_routes = {}
-    local uri_router
     local match_opts = {}
 function _M.match(api_ctx)
     local user_routes = _M.user_routes
```
Comment (kingluo) [inline apisix/http/router/radixtree_uri.lua:28]: should be removed too?

File: apisix/http/router/radixtree_uri.lua [hunk 6]
```diff
@@ -17,33 +17,15 @@
...
-    local _, service_version = get_services()
-    if not cached_router_version or cached_router_version ~= user_routes.conf_version
-        or not cached_service_version or cached_service_version ~= service_version
-    then
-        uri_router = base_router.create_radixtree_uri_router(user_routes.values,
```
[the remark this answers is not in the document]
Comment (ranxuxin001) [reply apisix/http/router/radixtree_uri.lua:37]: apisix/http/router/radixtree_host_uri.lua  is done.

**Pull request closed without merging**